IB Historical Bars Resubscription Fails After Connection Loss

# [BUG] IB Historical Bars Resubscription Fails After Connection Loss

## Confirmation

- [x] I have searched existing issues to confirm this bug has not already been reported
- [x] This bug was not present in an earlier version (v1.219.0 works correctly)
- [x] I can provide a detailed description of the bug
- [x] I can provide code snippets demonstrating the bug
- [x] I have included all relevant specifications

## Expected Behavior

After IB Gateway disconnects and reconnects (Error 1100 → Error 1102), historical bar subscriptions should automatically resubscribe and continue streaming bar data without interruption.

## Actual Behavior

After reconnection, resubscription silently fails with a `TypeError`, causing complete data starvation. No bar data flows despite the system appearing healthy and "successfully" reconnected.

**Production Impact**:
- From 7-hour live session: Only 17 bars received instead of expected ~420 (96% data loss)
- 24 disconnections occurred, 0 successful resubscriptions
- Zero trades executed due to strategies being starved for data

## Steps to Reproduce

1. Configure IB adapter with historical bars subscription:
   ```python
   config = InteractiveBrokersLiveDataClientConfig(
       ibg_host="127.0.0.1",
       ibg_port=7497,
       # ... other config
   )
   ```

2. Start trading node and verify bars are flowing

3. Wait for natural IB disconnect (happens every 4-30 minutes) OR manually restart IB Gateway

4. Observe reconnection messages (Error 1102)

5. **Result**: No more bar data is received, despite connection appearing successful

## Code Snippets or Logs

### Error Log
```
2025-09-29T18:46:07.541334000Z [ERROR] IB-PAPER-01.InteractiveBrokersClient-003:
Failed to resubscribe to Subscription(req_id=10006, name='AAPL.NASDAQ-1-MINUTE-LAST-EXTERNAL',
handle=functools.partial(..., start=1759170339583468000), ...)

TypeError(InteractiveBrokersClientMarketDataMixin.subscribe_historical_bars()
got an unexpected keyword argument 'start')
```

### Root Cause

**File**: `nautilus_trader/adapters/interactive_brokers/client/market_data.py`

**Problem at lines 457-466**:
```python
subscription = await self._subscribe(
    name,
    self.subscribe_historical_bars,
    self._eclient.cancelHistoricalData,
    bar_type=bar_type,
    contract=contract,
    use_rth=use_rth,
    handle_revised_bars=handle_revised_bars,
    start=start,  # ← BUG: Passed as separate kwarg
)
```

**Method signature at lines 416-422**:
```python
async def subscribe_historical_bars(
    self,
    bar_type: BarType,
    contract: IBContract,
    use_rth: bool,
    handle_revised_bars: bool,
    params: dict,  # ← Only accepts params dict, not start
) -> None:
```

**Issue**: PR #2870 introduced a parameter mismatch:
1. `start` is extracted from `params.get("start_ns")`
2. But then passed as separate `start=start` kwarg to `_subscribe()`
3. This creates a `functools.partial` with `start` in kwargs
4. On resubscription, the partial calls the method with `start` kwarg
5. Method signature only accepts `params` dict, causing `TypeError`
6. Exception is caught and logged, resubscription silently fails

### Proposed Fix

Store `start` times separately instead of in subscription handle:

```python
# 1. Add instance variable in __init__ (client.py)
self._subscription_start_times: dict[int, int] = {}

# 2. Fix subscription creation (market_data.py lines 457-466)
subscription = await self._subscribe(
    name,
    self.subscribe_historical_bars,
    self._eclient.cancelHistoricalData,
    bar_type=bar_type,
    contract=contract,
    use_rth=use_rth,
    handle_revised_bars=handle_revised_bars,
    params=params,  # ← Pass full params dict
)

# Store start separately for bar filtering
self._subscription_start_times[subscription.req_id] = start

# 3. Update bar filtering (line 900-902)
start = self._subscription_start_times.get(req_id)

# 4. Add cleanup in unsubscribe_historical_bars
subscription = self._subscriptions.get(name=name)
if subscription:
    self._subscription_start_times.pop(subscription.req_id, None)
```

**Files Modified**:
- `nautilus_trader/adapters/interactive_brokers/client/client.py` (+1 line)
- `nautilus_trader/adapters/interactive_brokers/client/market_data.py` (+14 lines, -4 lines)

## Specifications

- **Operating System**: macOS (also affects Linux/Windows)
- **Python Version**: 3.11
- **nautilus_trader Version**: v1.221.0 (git commit 7f52d2fb)
- **Affected Versions**: v1.220.0, v1.221.0+
- **Working Version**: v1.219.0 and earlier
- **Introduced in**: PR #2870
- **Broker**: Interactive Brokers Paper Trading (port 7497)

## Workaround

Downgrade to v1.219.0:
```toml
"nautilus-trader==1.219.0"
```

Trade-off: Lose features from v1.220.0+ (BitMEX support, funding rates, etc.)

## Additional Context

- IB Gateway frequent disconnects are **normal behavior**, not a broker bug
- Resubscription mechanism is **critical** for IB live trading reliability
- Bug creates **silent failure** - system appears healthy but is completely non-functional
- This is a **production-blocking issue** for anyone using IB adapter with live/paper trading and historical bars

## Fix Available

I have a working fix implemented on branch `fix/ib-historical-bars-resubscription-2870`. Happy to submit PR if needed.

--- a/nautilus_trader/adapters/interactive_brokers/client/client.py
+++ b/nautilus_trader/adapters/interactive_brokers/client/client.py
@@ -147,6 +147,7 @@ class InteractiveBrokersClient(
             {}
         )  # Track timeout tasks for each bar type
         self._subscription_tick_data: dict[int, dict] = {}  # Store tick data by req_id
+        self._subscription_start_times: dict[int, int] = {}  # Store start_ns for bar filtering
 
         # OrderMixin
         self._exec_id_details: dict[

--- a/nautilus_trader/adapters/interactive_brokers/client/market_data.py
+++ b/nautilus_trader/adapters/interactive_brokers/client/market_data.py
@@ -442,7 +442,7 @@ class InteractiveBrokersClientMarketDataMixin(BaseMixin):
         """
         name = str(bar_type)
         now = self._clock.timestamp_ns()
-        start = params.get("start_ns")
+        start = params.pop("start_ns", None)
 
         if start is not None:
             # start_time = pd.Timestamp(start)

--- a/nautilus_trader/adapters/interactive_brokers/client/market_data.py
+++ b/nautilus_trader/adapters/interactive_brokers/client/market_data.py
@@ -76,6 +76,7 @@ class InteractiveBrokersClientMarketDataMixin(BaseMixin):
 
     # Instance variables that will be available when mixed into InteractiveBrokersClient
     _subscription_tick_data: dict[int, dict[int, Any]]
+    _subscription_start_times: dict[int, int]  # reqId -> start_ns (for bar filtering)
 
     _order_books: ClassVar[dict[int, dict[str, dict[int, IBKRBookLevel]]]] = {}
     """
@@ -462,8 +463,11 @@ class InteractiveBrokersClientMarketDataMixin(BaseMixin):
             contract=contract,
             use_rth=use_rth,
             handle_revised_bars=handle_revised_bars,
-            start=start,
+            params=params,
         )
+
+        # Store start time separately for bar filtering (not part of resubscription handle)
+        self._subscription_start_times[subscription.req_id] = start
 
         bar_size_setting: str = bar_spec_to_bar_size(bar_type.spec)
         self._eclient.reqHistoricalData(
@@ -490,6 +494,12 @@ class InteractiveBrokersClientMarketDataMixin(BaseMixin):
 
         """
         name = str(bar_type)
+
+        # Clean up stored start time before unsubscribing
+        subscription = self._subscriptions.get(name=name)
+        if subscription:
+            self._subscription_start_times.pop(subscription.req_id, None)
+
         await self._unsubscribe(name, self._eclient.cancelHistoricalData)
 
     async def get_historical_bars(
@@ -893,9 +903,8 @@ class InteractiveBrokersClientMarketDataMixin(BaseMixin):
             if bar:
                 request.result.append(bar)
         elif subscription := self._subscriptions.get(req_id=req_id):
-            start = None
-            if isinstance(subscription.handle, functools.partial):
-                start = subscription.handle.keywords.get("start")
+            # Get start time from stored subscription start times
+            start = self._subscription_start_times.get(req_id)
 
             bar = await self._process_bar_data(
                 bar_type_str=str(subscription.name),

